engineapi/proxy.go

```diff
@@ -172,62 +425,241 @@ type Proxy struct {
 
 	grpcClient *imclient.ProxyClient
 
-	lock *sync.RWMutex
+	lock sync.RWMutex
+
+	inCall     int
+	inCallLock sync.RWMutex
+
+	nextResync time.Time
+	resyncLock sync.RWMutex
+
+	nextGC  *time.Time
+	gcRetry int
+	gcLock  sync.RWMutex
+}
+
+type EngineClientProxy interface {
+	EngineClient
+	BackupTargetBinaryClient
+
+	getEngineClientProxy() engineClientProxy
+
+	calling()
+	Done()
+}
+
+func (p *Proxy) getEngineClientProxy() engineClientProxy {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	return p
+}
+
+// calling increments the number of an engine proxy client concurrent calls.
+func (p *Proxy) calling() {
+	p.setInCall(true)
+}
+
+// Done decrements the number of an engine proxy client concurrent calls.
+func (p *Proxy) Done() {
+	p.setInCall(false)
+}
+
+type engineClientProxy interface {
+	getLogger() logrus.FieldLogger
+
+	isGRPC() bool
+	getConnectionState() connectivity.State
+
+	start(*longhorn.InstanceManager, logrus.FieldLogger, *datastore.DataStore, chan struct{}) error
+	stop() error
+
+	shouldCreate(*longhorn.InstanceManager) bool
+	shouldDelete() bool
```

Comment:
I feel you can make delete and gc consistent. You can use delete or either gc, but just choose one term, that will make the code understandable. WDYT?